### PR TITLE
Implemented system collections

### LINKED ISSUE #1590: System Collection
Each bundle should be able to create a "System Collection" to put there assets. for example the "avatars" in contact bundle.

### PR DESCRIPTION
**Pull request by wachterjohannes**
__tasks:__

- [x] test coverage
- [ ] gather feedback for my changes
- [x] submit changes to the documentation

__Related PR:__

* https://github.com/sulu-io/sulu-docs/pull/128
* https://github.com/massiveart/husky/pull/573
* https://github.com/sulu-io/sulu-standard/pull/547

__informations:__

| q                | a
| ---------------- | ---
| Fixed tickets    |  fixes #1590 
| BC breaks        | see upgrade
| Documentation PR | https://github.com/sulu-io/sulu-docs/pull/128

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -2,6 +2,23 @@
 
 ## dev-develop
 
+### System Collections
+Remove the config `sulu_contact.form.avatar_collection` and use it in the sql bellow as `{old-avatar-collection}`.
```
Comment (danrot) [inline UPGRADE.md]: `below`

File: UPGRADE.md [hunk 2]
```diff
@@ -2,6 +2,23 @@
...
+if you have not configured that the value for the placeholder is `1`. The value of `{new-system-collection-id}` you will
+get over the URL in the backend.
```
Comment (danrot) [inline UPGRADE.md]: These sentences confuse me.

File: src/Sulu/Bundle/ContactBundle/Resources/config/services.xml [hunk 3]
```diff
@@ -5,7 +5,7 @@
         <parameter key="sulu_contact.admin.class">Sulu\Bundle\ContactBundle\Admin\ContactAdmin</parameter>
         <parameter key="sulu_contact.contact_content_navigation_provider.class">Sulu\Bundle\ContactBundle\Admin\ContactContentNavigationProvider</parameter>
         <parameter key="sulu_contact.account_content_navigation_provider.class">Sulu\Bundle\ContactBundle\Admin\AccountContentNavigationProvider</parameter>
-        <parameter key="sulu_contact.js_config.class">Sulu\Bundle\AdminBundle\Admin\JsConfig</parameter>
+        <parameter key="sulu_contact.js_config.class">Sulu\Bundle\ContactBundle\Admin\JsConfigAvatarCollection</parameter>
```
Comment (danrot) [inline src/Sulu/Bundle/ContactBundle/Resources/config/services.xml]: I'd see that as a chance to get rid of the parameter, since it is not recommended in symfony 3.0 anymore.

File: src/Sulu/Bundle/MediaBundle/DataFixtures/ORM/LoadCollectionTypes.php [hunk 4]
```diff
@@ -21,20 +21,39 @@ class LoadCollectionTypes extends AbstractFixture implements OrderedFixtureInter
...
-
         // force id = 1
-        $metadata = $manager->getClassMetaData(get_class($collectionType));
+        $metadata = $manager->getClassMetaData(CollectionType::class);
         $metadata->setIdGeneratorType(\Doctrine\ORM\Mapping\ClassMetadata::GENERATOR_TYPE_NONE);
 
-        $collectionType->setName('collection.default');
-        $manager->persist($collectionType);
+        $defaultCollectionType = $this->createCollectionType(1, 'collection.default', 'Default');
+        $manager->persist($defaultCollectionType);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/DataFixtures/ORM/LoadCollectionTypes.php]: Why do we need a default collection?

File: src/Sulu/Bundle/MediaBundle/Entity/BaseCollection.php [hunk 5]
```diff
@@ -122,16 +123,6 @@ public function getCreator()
     }
 
     /**
-     * Constructor.
-     */
-    public function __construct()
-    {
-        $this->meta = new \Doctrine\Common\Collections\ArrayCollection();
-        $this->media = new \Doctrine\Common\Collections\ArrayCollection();
-        $this->children = new \Doctrine\Common\Collections\ArrayCollection();
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/BaseCollection.php]: Wouldn't it be better to let these initializations in this class? The it would work for all child classes (assuming they are using `parent::__construct()`)

File: src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php [hunk 6]
```diff
@@ -233,4 +234,24 @@ public function findCollectionBreadcrumbById($id)
...
+            ->leftJoin('collection.meta', 'collectionMeta')
+            ->leftJoin('collection.defaultMeta', 'defaultMeta')
+            ->where('collection.key = :key');
+
+        $query = $queryBuilder->getQuery();
+        $query->setParameter('key', $key);
+
+        try {
+            return $query->getSingleResult();
+        } catch (UnexpectedResultException $ex) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php]: When does this happen?

Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php]: Actually I think checking for the `NoResultException` would be better, because more specific.

File: src/Sulu/Component/Cache/ConfigCache.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Cache;
+
+use Symfony\Component\Config\ConfigCache as BaseConfigCache;
+
+class ConfigCache extends BaseConfigCache
```
Comment (danrot) [inline src/Sulu/Component/Cache/ConfigCache.php]: Why was that necessary?

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionBuilder.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+namespace Sulu\Component\Media\SystemCollections;
+
+use Massive\Bundle\BuildBundle\Build\BuilderContext;
+use Massive\Bundle\BuildBundle\Build\BuilderInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\DependencyInjection\ContainerAwareInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Warm-up cache for system collections.
```
Comment (danrot) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionBuilder.php]: Not a warm-up anymore, is it?

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        foreach ($this->config as $namespaceKey => $namespaceItem) {
+            $namespace = $this->getOrCreateCollection(
+                $namespaceKey,
+                $namespaceItem['meta_title'],
+                $userId,
+                $root->getId()
+            );
+            $collections[$namespaceKey] = $namespace->getId();
+
+            foreach ($namespaceItem['collections'] as $collectionKey => $collectionItem) {
```
Comment (danrot) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php]: Why are here two loops? Shouldn't that be called somehow recursive, to also support a deeper level of nesting?

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+    private function getOrCreateCollection($key, $localizedTitles, $userId, $parent)
+    {
+        $locales = array_keys($localizedTitles);
+        $firstLocale = array_shift($locales);
+
+        $collection = $this->collectionManager->getByKey($key, $firstLocale);
+        if ($collection === null) {
+            $collection = $this->createCollection($localizedTitles[$firstLocale], $key, $firstLocale, $userId, $parent);
+        } else {
+            $collection->setTitle($localizedTitles[$firstLocale]);
```
Comment (danrot) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php]: Just an idea: Wouldn't it be better to throw an exception here, instead of just updating a collection you do not know anything about?

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionManagerInterface.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+     * Returns id of system collection with given key.
+     *
+     * @param string $key
+     *
+     * @return int id of system collection
+     */
+    public function getSystemCollection($key);
+
+    /**
+     * Returns TRUE if given id is registered as system collection.
```
Comment (danrot) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionManagerInterface.php]: `true`

File: src/Sulu/Component/Media/SystemCollections/UnrecognizedSystemCollection.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    private $key;
+
+    /**
+     * @var string []
+     */
+    private $recognizedSystemCollections;
+
+    public function __construct($key, array $recognizedSystemCollections)
+    {
+        parent::__construct(sprintf('Unrecognized system collection "%s"', $key));
```
Comment (danrot) [inline src/Sulu/Component/Media/SystemCollections/UnrecognizedSystemCollection.php]: Maybe also add the known system collections in this error message.

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/CollectionControllerTest.php [hunk 13]
```diff
@@ -586,10 +629,14 @@ public function testPut()
 
         $this->assertNotEmpty($response);
 
-        $this->assertEquals(1, $response->total);
+        $this->assertEquals(2, $response->total); // 1 created 1 system collection
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/CollectionControllerTest.php]: I think we have to find a way around this... If another bundle adds more system collections, we also have to update these tests...

File: src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php (hunk 6 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php]: the `UnexpectedResultException` is the base class of `NonUniqueResultException` and `NoResultException` so it would catch both. should i change it anyway?

File: src/Sulu/Component/Cache/ConfigCache.php (hunk 7 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Cache/ConfigCache.php]: when i use the normal `ConfigCache` i cannot get the content from that class. then i have to inject the `$cache` variable and the path to the cache file ... but i will add a phpdoc.

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionBuilder.php (hunk 8 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionBuilder.php]: right wrong comment

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php (hunk 9 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php]: hmm the config is no able to go deeper but your right the service should be able to ... i will change that

File: src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php (hunk 10 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Media/SystemCollections/SystemCollectionManager.php]: but when we change the title in our bundles (or someone else) has to provide an upgrade script ... with this little line we have to do nothing ...

but i know something about ... i know that this collection is linked to this configuration with the key. an exception here is no solution because we need this to find the id of the systemcollection.

File: src/Sulu/Component/Media/SystemCollections/UnrecognizedSystemCollection.php (hunk 12 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Media/SystemCollections/UnrecognizedSystemCollection.php]: i have thought about this but i think the amount of system collections will be a lot so i think we should not add this into the messages

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/CollectionControllerTest.php (hunk 13 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/CollectionControllerTest.php]: thats right ... maybe fetch the configuration and add the amount here?

File: src/Sulu/Bundle/MediaBundle/Entity/BaseCollection.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/BaseCollection.php]: but these variables are not existing in this class ... they are added in the subclass!

File: src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php (hunk 6 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/CollectionRepository.php]: Do you really want to silently return null for the `NonUniqueResultException`? That doesn't feel right to me, because that would be a serious violation of database constraints, for the `NoResultException` that seems right.

File: src/Sulu/Bundle/MediaBundle/Entity/BaseCollection.php (hunk 5 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/BaseCollection.php]: ok, then it's fine

File: src/Sulu/Component/Media/SystemCollections/UnrecognizedSystemCollection.php (hunk 12 above)
Comment (danrot) [inline src/Sulu/Component/Media/SystemCollections/UnrecognizedSystemCollection.php]: Do you really think that well be that lot? I can't really imagine that...

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/CollectionControllerTest.php (hunk 13 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/CollectionControllerTest.php]: Isn't it possible to change the configuration in a way that it doesn't create any system collections for this test?

File: src/Sulu/Bundle/MediaBundle/Resources/config/services.xml [hunk 14]
```diff
@@ -171,5 +171,23 @@
         <service id="sulu_media.security_context" class="Sulu\Component\Security\Authorization\SecurityCondition">
             <argument>sulu.media.collections</argument>
         </service>
+
+        <service id="sulu_media.system_collections.cache" class="Sulu\Component\Cache\ConfigCache">
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/config/services.xml]: Why have you defined that as a service? Without that it would not be necessary to have our own `ConfigCache` class...

**Comment by wachterjohannes** [comment]
@danrot fixed

**Comment by danrot** [comment]
@wachterjohannes There are still open comments :wink:

File: src/Sulu/Bundle/MediaBundle/Resources/config/services.xml (hunk 14 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Resources/config/services.xml]: i will remove the own config cache class but the dependency injection i think is the better way. because then we can test the cache handling properly ... to load the file content i will use the `__toString` function.

File: UPGRADE.md [hunk 15]
```diff
@@ -2,6 +2,25 @@
...
+Update the database schema and then update the data-fixtures by running following sql statement.
+
+```bash
+app/console doctrine:schema:update --force
+```
+
+```sql
+UPDATE me_collection_types SET collection_type_key='collection.default', name='Default' WHERE id=1;
+INSERT INTO me_collection_types (id, name, collection_type_key) VALUES ('2', 'System Collections', 'collection.system');
+UPDATE me_media SET idCollections={new-system-collection-id} WHERE idCollections={old-avatar-collection};
```
Comment (danrot) [inline UPGRADE.md]: This line should be executed after the site was reloaded, otherwise the collection does not exist yet.
